python/ql/src/experimental/Security/CWE-022bis/UnsafeUnpack.ql

```diff
@@ -0,0 +1,136 @@
+
+  override predicate isSource(DataFlow::Node source) {
+    // A source coming from a remote location
+    exists(Http::Client::Request request | source = request)
+    or
+    //A source coming from a CLI argparse module
+    exists(Node o, API::Node ap, MethodCallNode args |
+      ap = API::moduleImport("argparse").getMember("ArgumentParser").getACall().getReturn() and
+      args = ap.getMember("parse_args").getACall() and
+      args.flowsTo(o) and
+      source.(AttrRead).accesses(o, any(string s))
+    )
+    or
+    // A source catching a S3 filename download
+    exists(API::Node s3 | source = s3.getMember("download_file").getACall().getArg(2))
+    or
+    // A source download a file using wget
+    exists(MethodCallNode mcn |
+      mcn = API::moduleImport("wget").getMember("download").getACall() and source = mcn.getArg(1)
+    )
+    or
+    // catch the uploaded files as a source
+    exists(Subscript s, Attribute at |
+      at = s.getObject() and at.getAttr() = "FILES" and source.asExpr() = s
+    )
+    or
+    exists(Node obj, AttrRead ar |
+      ar.getAMethodCall("get").flowsTo(source) and
+      ar.accesses(obj, "FILES")
+    )
+    or
+    exists(Node obj, AttrRead ar |
+      ar.getAMethodCall("getlist").flowsTo(source) and
+      ar.accesses(obj, "FILES")
+    )
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    // A sink capturing method calls to `unpack_archive`.
+    sink = API::moduleImport("shutil").getMember("unpack_archive").getACall().getArg(0)
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    // Writing the response data to the archive
+    exists(Stdlib::FileLikeObject::InstanceSource is, Node f, MethodCallNode mc |
+      is.flowsTo(f) and
+      mc.getMethodName() = "write" and
+      f = mc.getObject() and
+      nodeFrom = mc.getArg(0) and
+      nodeTo = is.(CallCfgNode).getArg(0)
+    )
+    or
+    // Copying the response data to the archive
+    exists(Stdlib::FileLikeObject::InstanceSource is, Node f, MethodCallNode mc |
+      is.flowsTo(f) and
+      mc = API::moduleImport("shutil").getMember("copyfileobj").getACall() and
+      f = mc.getArg(1) and
+      nodeFrom = mc.getArg(0) and
+      nodeTo = is.(CallCfgNode).getArg(0)
+    )
+    or
+    // Reading the response
+    exists(MethodCallNode mc |
+      nodeFrom = mc.getObject() and
+      mc.getMethodName() = "read" and
+      mc.flowsTo(nodeTo)
+    )
+    or
+    // Accessing the name or raw content
+    exists(AttrRead ar | ar.accesses(nodeFrom, ["name", "raw"]) and ar.flowsTo(nodeTo))
+    or
+    //Use of join of filename
+    exists(API::CallNode mcn |
+      mcn = API::moduleImport("os").getMember("path").getMember("join").getACall() and
+      nodeFrom = mcn.getArg(1) and
+      mcn.flowsTo(nodeTo)
+    )
+    or
+    // Read by chunks
+    exists(MethodCallNode mc |
+      nodeFrom = mc.getObject() and mc.getMethodName() = "chunks" and mc.flowsTo(nodeTo)
+    )
```

Comment:
What is the api? Can we constrain `nodeFrom` to it?
Also, if it is filenames that we a tracking, is this step right?

Reply:
This is related to Django API in [here](https://docs.djangoproject.com/en/4.1/ref/files/uploads/#django.core.files.uploadedfile.UploadedFile.chunks).

Reply:
I think, this is better:
```codeql
nodeTo.(MethodCallNode).calls(nodeFrom, "chunks")
```

Reply:
We are tracking `request.Files[]`